### PR TITLE
BGL: Connected_component_graph and copy_face_graph_patch

### LINKED ISSUE #1802: Add a function to copy a patch
We have  the function `copy_face_graph()`, it would be interesting to have a function `copy_face_graph_patch()`. The API of the function should match the various overloads of `connected_components()` (using cc_id, using a face seed,...).

@afabri mentioned that more generally we may have an adapter that make a connected component a read-only model of `face_graph`. We first need to see if this is useful for other applications.

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

This PR adds a wrapper that allows to use a patch as a whole graph and a function to copy only a patch of a graph, based on connected_components().

## Release Management

* Affected package(s):BGL
* Issue(s) solved (if any): fix #1802 
* Feature/Small Feature (if any): [`Features/Small_Features/BGL::Connected_components_graph`](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/BGL::Connected_components_graph)

### CODE REVIEW

File: BGL/include/CGAL/boost/graph/copy_face_graph_patch.h [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+                           Src_vpm sm_vpm, Tgt_vpm tm_vpm )
+#endif
+{
+
+    typedef typename boost::graph_traits<SourceMesh>::face_descriptor g_face_descriptor;
+    typedef typename boost::graph_traits<SourceMesh>::faces_size_type faces_size_t;
+    typedef boost::associative_property_map< std::map< g_face_descriptor, faces_size_t>  >FCMap;
+    typedef CGAL::Connected_component_graph<SourceMesh, FCMap> Adapter;
+
+    std::map<g_face_descriptor, faces_size_t> map;
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/copy_face_graph_patch.h]: `boost::unordered_map map(num_faces(sm))` ?

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+        : _graph(graph), property_map(fccmap), patch_index(pid)
+    {
+    }
+
+
+    Connected_component_graph(const Connected_component_graph&f)
+        : _graph(f.graph()), property_map(f.propertyMap()), patch_index(f.patchIndex())
+    {
+    }
+
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: Remove it, the default is perfectly suited.

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+        Is_simplex_valid(const Self* graph)
+            :adapter(graph)
+        {}
+
+        Is_simplex_valid()
+            :adapter(NULL)
+        {}
+        template<typename Simplex>
+        bool operator()(Simplex s)
+        {
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: add an assertion checking `graph!=NULL`

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+    return num_halfedges(w.graph());
+}
+
+// FaceGraph
+template <class Graph, typename FaceComponentMap >
+typename boost::graph_traits< Connected_component_graph<Graph, FaceComponentMap> >::face_descriptor
+face(typename boost::graph_traits< Connected_component_graph<Graph, FaceComponentMap> >::halfedge_descriptor h,
+     const Connected_component_graph<Graph, FaceComponentMap> & w)
+{
+    CGAL_assertion(CGAL::is_valid(h, w));
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: if `h` is a border halfedge (on the border of the patch pointing outside the patch), `null_face()` should be returned.

File: BGL/include/CGAL/boost/graph/copy_face_graph_patch.h [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                           Src_vpm sm_vpm, Tgt_vpm tm_vpm )
+#endif
+{
+
+    typedef typename boost::graph_traits<SourceMesh>::face_descriptor g_face_descriptor;
+    typedef typename boost::graph_traits<SourceMesh>::faces_size_type faces_size_t;
+    typedef boost::associative_property_map< boost::unordered_map< g_face_descriptor, faces_size_t>  >FCMap;
+    typedef CGAL::Connected_component_graph<SourceMesh, FCMap> Adapter;
+
+    boost::unordered_map<g_face_descriptor, faces_size_t> map(CGAL::num_faces(sm);
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/copy_face_graph_patch.h]: missing `)`

**Comment by afabri** [comment]
I am not sure that we need a function `copy_face_graph_patch()`.  The `Connected_component_graph` should be enough (in fact the function is only a 2 liner).

File: BGL/test/BGL/graph_concept_Connected_component_graph.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
+#include <CGAL/boost/graph/graph_traits_Surface_mesh.h>
```
Comment (afabri) [inline BGL/test/BGL/graph_concept_Connected_component_graph.cpp]: Just include Surface_mesh.h and not the graph traits and properties files.

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+//
+// Author(s)     : Maxime Gimeno
+#ifndef CGAL_BOOST_GRAPH_CONNECTED_COMPONENT_GRAPH_H
+#define CGAL_BOOST_GRAPH_CONNECTED_COMPONENT_GRAPH_H
+
+#include <CGAL/assertions.h>
+#include <CGAL/boost/graph/properties.h>
+#include <CGAL/boost/graph/iterator.h>
+#include <CGAL/assertions.h>
+#include <CGAL/Polygon_mesh_processing/connected_components.h>
```
Comment (afabri) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: Should not be needed.

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+    typedef typename BGTG::traversal_category traversal_category;
+    typedef typename BGTG::vertices_size_type vertices_size_type;
+    typedef typename BGTG::edges_size_type edges_size_type;
+    typedef typename BGTG::halfedges_size_type halfedges_size_type;
+    typedef typename BGTG::faces_size_type faces_size_type;
+    typedef typename BGTG::degree_size_type degree_size_type;
+
+    static vertex_descriptor null_vertex()
+    {
+        return vertex_descriptor(BGTG::null_vertex());
```
Comment (afabri) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: No need for `vertex_descriptor`

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+                     FaceComponentMap fccmap,
+                     typename boost::property_traits<FaceComponentMap>::value_type pid)
+        : _graph(graph), property_map(fccmap), patch_index(pid)
+    {
+    }
+
+
+    const Graph& graph()const{ return _graph; }
+
+    FaceComponentMap propertyMap()const{ return property_map; }
```
Comment (afabri) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: property_map()

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+{
+    CGAL_assertion(is_valid(u, w) && is_valid(v, w));
+    typename boost::graph_traits<Connected_component_graph<Graph, FaceComponentMap> >::edge_descriptor e = edge(u, v, w.graph()).first;
+    bool res = is_valid(e, w);
+    return std::make_pair(e, res);
+}
+
+
+template <class Graph, typename FaceComponentMap >
+std::pair<typename boost::graph_traits<Connected_component_graph<Graph, FaceComponentMap> >::vertex_iterator,
```
Comment (afabri) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: Replace the `std::par` with a `CGAL::Iterator_range`. 
Please do this also in the `Gwdwg`.

File: BGL/include/CGAL/boost/graph/Connected_component_graph.h [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+bool
+is_valid(const typename boost::graph_traits< Connected_component_graph<Graph, FaceComponentMap> >::face_descriptor f,
+         const Connected_component_graph<Graph, FaceComponentMap> & w)
+{
+    return boost::get(w.propertyMap(), f) == w.patchIndex();
+}
+
+template <class Graph, typename FaceComponentMap >
+bool
+is_valid(const typename boost::graph_traits< Connected_component_graph<Graph, FaceComponentMap> >::halfedge_descriptor h,
```
Comment (afabri) [inline BGL/include/CGAL/boost/graph/Connected_component_graph.h]: `in_CC`  or in_connected_component` would probably be a better name.

**Comment by afabri** [comment]
The API of  `Connected_component_graph` should also have a function that allows to change the ID.

**Comment by sloriot** [comment]
About the naming: do we have any practical restriction to limit this class to only one connected component?  If not, then we better talk about subgraph or patches. Having the complement might also be interesting.

**Comment by sloriot** [comment]
About the copy function, I think you are right. Even the version that uses the seed face is smalI (and removing it solves  the pb of BGL using a PMP function)

**Comment by lrineau** [comment]
@maxGimeno Sébastien asked a question: https://github.com/CGAL/cgal/pull/1861#issuecomment-275978226

This PR is not waiting for review. It is waiting for your answers after the feedback from Sébastien.

And once you somehow both agree (you and @sloriot and @afabri), then you must submit a small feature, and a documentation.

**Comment by maxGimeno** [comment]
@sloriot We could use a std::set for passing the ids instead of a simple int, or any other data structure that provides a simple way to test if a value is contained or not. Maybe make another overload of the function for the single patch use case ?

**Comment by maxGimeno** [comment]
I removed the copy_face_graph_patch and used a boost::unordered_set of patch_ids as argument to pass several ids instead of one. I also added a constructor taking a single id, and adapted the tests to test with 2 patches instead of one.

**Comment by maxGimeno** [comment]
[Small feature page](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/BGL::Connected_components_graph)

File: BGL/doc/BGL/PackageDescription.txt [hunk 12]
```diff
@@ -157,6 +158,7 @@ user might encounter.
 
 - `CGAL::clear()`
 - `CGAL::copy_face_graph()`
+- `CGAL::copy_face_graph_patch()`
```
Comment (lrineau) [inline BGL/doc/BGL/PackageDescription.txt]: It seems `CGAL::copy_face_graph_patch()` is no longer is the PR.

File: BGL/include/CGAL/boost/graph/Connected_components_graph.h [hunk 13]
```diff
@@ -0,0 +1,620 @@
...
+
+    /*!
+     * \brief Creates a Connected_components_graph of the connected components of `graph` that are listed in `pids`.
+     * \param graph the graph containing the wanted patches.
+     * \param fccmap the property_map that assigns a patch to each face
+     * \param pids the indices of the patches of interest.
+     */
+    Connected_components_graph(const Graph& graph,
+                     FaceComponentMap fccmap,
+                     boost::unordered_set<typename boost::property_traits<FaceComponentMap>::value_type> pids)
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/Connected_components_graph.h]: you should avoid copying `pids`

File: BGL/include/CGAL/boost/graph/Connected_components_graph.h [hunk 14]
```diff
@@ -0,0 +1,620 @@
...
+        : _graph(graph), _property_map(fccmap), _patch_indices(pids)
+    {
+    }
+
+    /*!
+     * \brief Creates a Connected_components_graph of the connected component `pid` of `graph`.
+     * \param graph the graph containing the wanted patch.
+     * \param fccmap the property_map that assigns a patch to each face
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/Connected_components_graph.h]: key_type, value_type, category?

**Comment by lrineau** [comment]
@CGAL/geometryfactory The small feature [`Features/Small_Features/BGL::Connected_components_graph`][1] has been submitted two months ago. Now that we are working for CGAL-4.11, we should find time to review it.

  [1]: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/BGL::Connected_components_graph

**Comment by maxGimeno** [comment]
Great ! I'm gonna need that for #1931

**Comment by sloriot** [comment]
I'd like to see an alternative implementation using index map and bitset. The bitset are filled when the graph is constructed (or updated if it matters). I expect it to be much faster. I don't think the extra storage should be a problem, even for large meshes. Another advantage is also a tight `num_vertices()`, `num_faces()`... that is useful for `reserve()` in case for example of a call to `copy_face_graph()`.
Some benchmark between the two implementation would be welcome to decide whether one or the two implementation should be kept (which would have some consequences on the API).

**Comment by maxGimeno** [comment]
Here is a benchmark of the worst case scenario (when the connected_components_graph is the original graph) : 
bench1
bench2
bench3
bench4

With those results, is it reasonable to say that we should replace the original implementation with the one base don bitsets >

**Comment by lrineau** [comment]
Please solve the conflicts with `master`, and then I will merge the PR into `integration`.

**Comment by maxGimeno** [comment]
Apparently there are some problems under Windows with OpenMesh in the BGL tests of this branch, but I don't see why as I don't need or use OpenMesh...

**Comment by sloriot** [comment]
There is an extra typename error not related to OpenMesh. About the link pb, I guess that some header files from OpenMesh are included in some files that should not.

**Comment by sloriot** [comment]
Your last commit is only hiding the issue IMO. If you have a link error while you did not ask for OpenMesh then it means some header files were included but they should not.

**Comment by lrineau** [comment]
`test_Prefix.h`, from `BGL/test/BGL/` does include OpenMesh as soon as it is detected by CMake. That means that all test files using that header must be linked with OpenMesh libraries.

**Comment by maxGimeno** [comment]
@lrineau and I added a test for openmesh as it was missing, regarding of what you explained to me.

**Comment by lrineau** [comment]
Does not compile locally:
```
/home/lrineau/Git/cgal/BGL/test/BGL/test_Face_filtered_graph.cpp: In instantiation of ‘void test_halfedge_around_vertex_iterator(const Graph&) [with Graph = OpenMesh::PolyMesh_ArrayKernelT<>]’:
/home/lrineau/Git/cgal/BGL/test/BGL/test_Face_filtered_graph.cpp:292:41:   required from ‘void test(const std::vector<T>&) [with Graph = OpenMesh::PolyMesh_ArrayKernelT<>]’
/home/lrineau/Git/cgal/BGL/test/BGL/test_Face_filtered_graph.cpp:402:20:   required from here
/home/lrineau/Git/cgal/BGL/test/BGL/test_Face_filtered_graph.cpp:20:75: error: no matching function for call to ‘num_faces(const OpenMesh::PolyMesh_ArrayKernelT<>&)’
   boost::unordered_map<g_face_descriptor, std::size_t> map(CGAL::num_faces(g));
                                                            ~~~~~~~~~~~~~~~^~~
```
Have you tried locally?

**Comment by afabri** [comment]
num_faces() is in general found via König lookup, so the function should be in the same namespace as the graph class.

**Comment by lrineau** [comment]
That seems fixed, now, in https://github.com/CGAL/cgal/pull/1861/commits/c81e4444bf9d7b5ed847b8cc6a98c54e0f54d407.

**Comment by lrineau** [comment]
There are Doxygen errors/warnings:
```
doc/Polygon_mesh_processing/examples.txt:13: Warning: included file Polygon_mesh_processing/filtered_face_graph_example.cpp is not found. Check your EXAMPLE_PATH
doc/Polygon_mesh_processing/Polygon_mesh_processing.txt:628: Warning: unable to resolve reference to `Polygon_mesh_processing/face_filtered_graph_example.cpp' for \ref command
```
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.11-Ic-31/Polygon_mesh_processing.log

**Comment by lrineau** [comment]
There is still [a warning, on Windows][*]:
```
[ 50%] Building CXX object CMakeFiles/extract_segmentation_into_mesh_example.dir/extract_segmentation_into_mesh_example.cpp.obj
	C:\PROGRA~2\MICROS~1\2017\PROFES~1\VC\Tools\MSVC\1410~1.249\bin\HostX64\x64\cl.exe  @C:\cygwin\tmp\nm3F0E.tmp
extract_segmentation_into_mesh_example.cpp
Unknown compiler version - please run the configure tests and report the results
C:\CGAL\test\CGAL-4.11-Ic-34\include\CGAL/boost/graph/Face_filtered_graph.h(394): warning C4267: 'return': conversion from 'size_t' to 'unsigned int', possible loss of data
C:\CGAL\test\CGAL-4.11-Ic-34\include\CGAL/boost/graph/Face_filtered_graph.h(393): note: while compiling class template member function 'unsigned int CGAL::Face_filtered_graph<SM,CGAL::SM_index_pmap<Point,CGAL::SM_Face_index>,CGAL::SM_index_pmap<Point,CGAL::SM_Vertex_index>,CGAL::SM_index_pmap<Point,CGAL::SM_Halfedge_index>>::number_of_halfedges(void) const'
        with
        [
            Point=CGAL::Point_3<CGAL::Epick>
        ]
C:\CGAL\test\CGAL-4.11-Ic-34\include\CGAL/boost/graph/Face_filtered_graph.h(988): note: see reference to function template instantiation 'unsigned int CGAL::Face_filtered_graph<SM,CGAL::SM_index_pmap<Point,CGAL::SM_Face_index>,CGAL::SM_index_pmap<Point,CGAL::SM_Vertex_index>,CGAL::SM_index_pmap<Point,CGAL::SM_Halfedge_index>>::number_of_halfedges(void) const' being compiled
        with
        [
            Point=CGAL::Point_3<CGAL::Epick>
        ]
C:\CGAL\test\CGAL-4.11-Ic-34\cmake\platforms\x64_Cygwin-Windows10_MSVC2017-Release-64bits\test\Surface_mesh_segmentation_Examples\extract_segmentation_into_mesh_example.cpp(43): note: see reference to class template instantiation 'CGAL::Face_filtered_graph<SM,CGAL::SM_index_pmap<Point,CGAL::SM_Face_index>,CGAL::SM_index_pmap<Point,CGAL::SM_Vertex_index>,CGAL::SM_index_pmap<Point,CGAL::SM_Halfedge_index>>' being compiled
        with
        [
            Point=CGAL::Point_3<CGAL::Epick>
        ]

```
  [*]: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-34/Surface_mesh_segmentation_Examples/TestReport_afabri_x64_Cygwin-Windows10_MSVC2017-Release-64bits.gz

**Comment by lrineau** [comment]
There is still a warning, hopefully the last one:
```
/home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/test/BGL/test_Face_filtered_graph.cpp:366:21: warning: unused variable 'v' [-Wunused-variable]
  vertex_descriptor v = source(h, fga);
                    ^
```
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-35/BGL/TestReport_lrineau_ArchLinux-clang-Release.gz

**Comment by MaelRL** [comment]
The function `face(hd, filtered_graph)` seems bugged if `hd` is a border halfedge: `face(hd, filtered_graph.graph())` is then the `null_face` of the underlying graph and the bitset will segfault in `in_CC(face(hd, filtered_graph.graph()))`.

I propose the following patch:

```
...
-  if(in_CC(face(h,w.graph()), w))
+
+  if(face(h, w.graph()) == boost::graph_traits<Graph>::null_face()) // h is a border hafedge
+    return boost::graph_traits< CGAL::Face_filtered_graph<Graph, FIMap, VIMap, HIMap> >::null_face();
+  else if(in_CC(face(h,w.graph()), w))
...
```

**Comment by maxGimeno** [comment]
done

File: BGL/test/BGL/test_Prefix.h [hunk 15]
```diff
@@ -149,7 +149,7 @@ std::vector<OMesh> omesh_data()
 template <typename Graph>
 struct Surface_fixture_1 {
   Surface_fixture_1() {
-    assert(read_a_mesh(m, "data/fixture1.off"));
+    assert(read_a_mesh(m, "/home/gimeno/CGAL/BGL/test/BGL/data/fixture1.off"));
     assert(CGAL::is_valid(m));
```
Comment (MaelRL) [inline BGL/test/BGL/test_Prefix.h]: Some foreign changes hitched a ride on the previous commit, apparently

Comment (maxGimeno) [inline BGL/test/BGL/test_Prefix.h]: oops

**Pull request merged**